README.md
xbee-api
========

This is a Java API for communication with XBee/XBee-Pro series 1 (802.15.4) and series 2 (ZB/ZigBee Pro) OEM RF Modules, in API mode.  The objective of this project is to provide a flexible and simple to use API to interact with XBee radios.  In terms of flexibility, the goal is to not constrain you to a particular implementation (e.g. GUI, Web App, Processing etc.), but instead let you decide how to use it.  This software has been tested on Windows, Mac, and Linux and can run on any other platform that supports Java 5 or greater and RXTX, including the [Rasberry Pi](http://www.amazon.com/gp/product/B009SQQF9C/ref=as_li_qf_sp_asin_tl?ie=UTF8&camp=1789&creative=9325&creativeASIN=B009SQQF9C&linkCode=as2&tag=xbapra-20)! If you're interested in interfacing XBees directly with Arduinos see [XBee Arduino](https://github.com/andrewrapp/xbee-arduino)

**Note: This software requires API mode, by setting  AP=2. If you are using Series 2 XBee, you'll need to install API Firmware (coordinator, router or endpoint) with X-CTU (windows only) since they ship with AT firmware.  This software will not work correctly with AP=1**  Refer to [XBeeConfiguration](XBeeConfiguration.md) and WhyApiMode for more info.

## News ##
  * 2/19/15 I've created [xbee-socket](https://github.com/andrewrapp/xbee-socket) and [xbee-serial-server](https://github.com/andrewrapp/xbee-serial-server) projects for sharing a single xbee radio with multiple applications, over sockets. This is very beta. Also, [this](https://github.com/andrewrapp/xbee-api) code is now on github.
  * 2/1/14 Converted repository from Subversion to Git
  * 6/11/13 Yeah, you can use [xbee-api with your Raspberry Pi](http://rapplogic.blogspot.com/2013/06/xbee-on-raspberry-pi.html)
  * 4/3/11 I have created a [XBeeUseCases](http://code.google.com/p/xbee-api/wiki/XBeeUseCases) wiki that describes several use cases for communicating with XBees.
  * 11/7/10 The 0.9 release is now available!  This release includes a lot of new features and an extensive redesign of the internals.  See ReleaseNotes for details.
  * 8/13/10 I've added a [wiki](Processing.md) that describes how to use XBee-API with Processing
  * 3/29/09 I have released [XBee Arduino](http://code.google.com/p/xbee-arduino/).  This is a port of XBee API to the Arduino platform, allowing you to send/receive packets on Arduino.
  * 4/5/08 Initial Release

## Documentation ##

Note: some image links may not render correctly from the googlecode export

  * [Developer's Guide](https://github.com/andrewrapp/xbee-api/blob/wiki/DevelopersGuide.md)
  * [Use Cases](https://github.com/andrewrapp/xbee-api/blob/wiki/XBeeUseCases.md)
  * [XBee Configuration](https://github.com/andrewrapp/xbee-api/blob/wiki/XBeeConfiguration.md)
  * [Choosing an XBee](https://github.com/andrewrapp/xbee-api/blob/wiki/ChoosingAnXBee.md)
  * [Why API Mode](https://github.com/andrewrapp/xbee-api/blob/wiki/WhyApiMode.md)
  * [XBee Pin Reference](https://github.com/andrewrapp/xbee-api/blob/wiki/XBeePins.md)
  * [Digi XBee Series 1/802.15.4/S1](http://www.digi.com/products/wireless-wired-embedded-solutions/zigbee-rf-modules/point-multipoint-rfmodules/xbee-series1-module#docs) [(Product Manual)](http://ftp1.digi.com/support/documentation/90000982_S.pdf)
  * [Digi XBee Series 2/ZB/S2](http://www.digi.com/products/wireless-wired-embedded-solutions/zigbee-rf-modules/zigbee-mesh-module/xbee-zb-module#docs) [(Product Manual)](http://ftp1.digi.com/support/documentation/90000976_V.pdf)
  * [All wiki docs ported from googlecode](https://github.com/andrewrapp/xbee-api/tree/wiki)

[Google Group](https://groups.google.com/forum/#!forum/xbee-api)

## Learning/Books ##
  * [Wireless Sensor Networks: with ZigBee, XBee, Arduino, and Processing](http://www.amazon.com/gp/product/0596807732?ie=UTF8&tag=xbapra-20&linkCode=as2&camp=1789&creative=9325&creativeASIN=0596807732Building) (Kindle version available)
  * [Making Things Talk: Using Sensors, Networks, and Arduino to see, hear, and feel your world](http://www.amazon.com/gp/product/1449392431/ref=as_li_ss_tl?ie=UTF8&camp=1789&creative=390957&creativeASIN=1449392431&linkCode=as2&tag=xbapra-20)

<a />

## Examples ##

Here are a few examples that demostrate some common functionality:

Turn on an I/O port on a remote Series 2 (ZNet/ZB Pro) XBee:

```
XBee xbee = new XBee();
xbee.open("COM5", 9600);

// this is the Serial High (SH) + Serial Low (SL) of the remote XBee			
XBeeAddress64 addr64 = new XBeeAddress64(0xa, 0xb, 0xc, 0xd, 0xe, 0xf, 0, 1);

// Turn on DIO0 (Pin 20)
RemoteAtRequest request = new RemoteAtRequest(addr64, "D0", new int[{XBeePin.Capability.DIGITAL_OUTPUT_HIGH.getValue()});

xbee.sendAsynchronous(request);

RemoteAtResponse response = (RemoteAtResponse) xbee.getResponse();

if (response.isOk()) {
    System.out.println("Successfully turned on DIO0");
} else {
    System.out.println("Attempt to turn on DIO0 failed.  Status: " + response.getStatus());
}

// shutdown the serial port and associated threads
xbee.close();

```

Receive I/O samples from a remote Series 1 (802.15.4) XBee:

```
XBee xbee = new XBee();		
xbee.open("COM15", 9600);
			
while (true) {
    RxResponseIoSample ioSample = (RxResponseIoSample) xbee.getResponse();

    System.out.println("We received a sample from " + ioSample.getSourceAddress());	
			
    if (ioSample.containsAnalog()) {
        System.out.println("10-bit temp reading (pin 19) is " + ioSample.getSamples()[0].getAnalog1();
    }
}
```

There are many more detailed examples included in the code for both ZigBee and 802.15.4.  Here are just a few:

This example contains a few sample configurations to get your ZigBee coordinator and end devices configured quickly, without needing X-CTU (windows only) [ZNetApiAtExample.java](http://code.google.com/p/xbee-api/source/browse/trunk/src/com/rapplogic/xbee/examples/zigbee/ZNetApiAtExample.java)

Example of sending packets from a ZigBee coordinator to an end device [ZNetSenderExample.java](http://code.google.com/p/xbee-api/source/browse/trunk/src/com/rapplogic/xbee/examples/zigbee/ZNetSenderExample.java)

This example receives the packets sent in the previous example [ZNetReceiverExample.java](http://code.google.com/p/xbee-api/source/browse/trunk/src/com/rapplogic/xbee/examples/zigbee/ZNetReceiverExample.java)

## Feature Support ##

The majority of the XBee specification has been implemented for both 802.15.4 (Series 1) and ZNet 2.5/ZB Pro (Series 2).  See FeatureSupport for a full listing of supported API types.

## About ##

I received my first pair of XBee Pro radios in Nov '07, after learning about them on the Arduino website.  I quickly learned that to take advantage of the real power in XBee radios (e.g. I/O Samples, ACK/Delivery Status, Remote AT etc.), you need to configure the radio in API mode.  I searched for open source XBee software with support for API mode, but couldn't find anything, so I decided to write my own.  I started coding on 12/15/2007 and after about 300 commits and several months later I released the software on Google Code.

## Support ##

Please report any bugs on the [Issue Tracker](http://code.google.com/p/xbee-api/issues/list)  For questions regarding this software, not covered in the documentation, use the [Forum](http://groups.google.com/group/xbee-api?pli=1)


## Questions/Feedback ##

Questions about this project should be posted to  http://groups.google.com/group/xbee-api?pli=1  Be sure to provide as much detail as possible (e.g. what radios s1 or s2, firmware versions, configuration and code).

## Consulting/Commercial Licenses ##

I'm available for consulting to assist businesses or entrepreneurs that need help getting their projects up and running. I can also provide a commercial license for situations where you need to distribute code to clients/third parties that would otherwise conflict with GPL. For these matters I can be contacted at andrew.rapp `[`at`]` gmail.


ReleaseNotes.md
# Release Notes #

Version 0.9 (11/7/10)

This release includes the following updates

  * Added protocol independent XBeeConnection interface.  This allows developers to develop support for communicating with the radio over protocols other than serial (e.g. socket, xmpp).  The implementation overrides the providerConnection method.
  * PacketListener is now handled through a thread pool.
  * Added ResponseQueueFilter to determine which packet types will be added to the response queue.
  * Created NoRequestResponse interface for tagging XBeeResponse classes that are send by the radio/network w/o a corresponding request.
  * Created XBeeConfiguration class for specifying config information
  * XBee no longer extends RXTX.  Additionally, you can call open and close methods to your hearts content without issues
  * Removed the ZNET prefix from ZNET\_REMOTE\_AT\_REQUEST as it is present in Series 1 now.  Same with ZNET\_REMOTE\_AT\_RESPONSE
  * Fixed bug with Series 1 NodeDiscover example
  * ZB I/O Samples: Added parseIsSample(AtCommandResponse response) for parsing an I/O sample via RemoteAT.  Added Boolean isDigitalOn(int pin),  Integer getAnalog(int pin), boolean isDigitalEnabled(int pin), and boolean isAnalogEnabled(int pin) for dynamic access to digital and analog pins
  * WPAN I/O Samples: Added parse(IIntArrayInputStream ps), Added Integer getAnalog(int pin) , Boolean isDigitalOn(int pin), boolean isDigitalEnabled(int pin), and boolean isAnalogEnabled(int pin) for dynamic access to digital and analog pins
  * Removed getLastResponse(), getResponseBlocking(int), getResponseBlocking(), waitForResponse(int), and waitForResponse() methods from XBee class. Use getResponse and PacketListener interface instead.
  * Added GenericResponse.  If an unknown packet is encountered, it will be parsed into a Generic Response and the packet data may be accessed. Previously an exception was thrown if an unknown packet was read
  * ZNetRemoteAtRequest was renamed to RemoteAtRequest; ZNetRemoteAtResponse renamed to RemoteAtResponse
  * Added a startup heuristic to determine radio type and if it's configured correctly
  * getResponse now uses a BlockingQueue.
  * added clearResponseQueue method to XBee class
  * added equals and hashCode implementation to many classes
  * added getRawPacketBytes() and getProcessedPacketBytes() methods to XBeeResponse
  * iosamples getAnalog(pin)
  * renamed       DEFAULT\_OPTION (0),DISABLE\_ACK\_OPTION (1),BROADCAST\_OPTION(4); to UNICAST (0),DISABLE\_ACK (1),BROADCAST(4);
  * removed       public boolean isError() from TxStatusResponse.  use !isSuccess() instead
  * added getAnalog(int pin) and isDigitalOn(int pin) methods to ZNetRxIoSampleResponse
  * ZNetTxRequest now has Option enum instead of int
  * added BroadcastSender/Receiver examples

Version 0.5.5 (5/10/09)

This release includes the following updates

  * Maximum payload size is no longer enforced for both series 1 and 2 XBees.  The reason for this change is the max size varies depending on the firmware.  For example ZNet is 72 bytes but ZB Pro is up to 84 bytes and less based on certain configurations such as encryption.   A new method (setMaxPayloadSize) was added to allow the user to define a max size.  An exception will be thrown during send if the payload size exceeds the user defined limit.
  * Renamed TX\_16\_STATUS\_RESPONSE to TX\_STATUS\_RESPONSE in ApiId to eliminate potential confusion as it also applies to 64 bit addressing
  * Added NodeDiscover class for Series 1 XBees
  * Added example of using NodeDiscover for series 2: NodeDiscoverTest.java
  * Added isSuccess() method to TxStatusResponse (series 1) and ZNetTxStatusResponse (series 2)
  * Added NO\_RESPONSE (4) status response enum that is new in series 1 10C8 firmware
  * Renamed MAX\_PAYLOAD\_SIZE to ZNET\_MAX\_PAYLOAD\_SIZE since it doesn't apply to ZB Pro firmware
  * Added ZbForceSample class.  Going forward I'll use the generic ZB (ZigBee) prefix instead of ZNet.  Except where noted, both ZNet and ZB Pro firmware apply to all ZNet and ZB classes
  * Fixed bug with node id.  Thanks to Nerio Montoya for reporting
  * Added INFO log level for packets going in/out of the serial.  this will make it easier to analyze traffic w/o the verbosity of the debug level
  * Added equals implementation for 64-bit address to resolve RuntimeException issue
  * XBee now stops parsing packets when serial cable is disconnected.  Previously RXTX would go bonkers and fill up your hard drive with log errors
  * Added isAddressBroadcast and isPanBroadcast methods for series 1 RX
  * Added new ZB PRO specific status to tx status response class
  * Replaced toChar with toString in byte utils
  * Added ZB Pro manual and updated other manuals.
  * Modified ZNetRemoteAtRequest to subclass AtCommand and ZNetRemoteAtResponse to subclass AtCommandResponse.   To perform this change it was necessary to change the return type of the getStatus method in AtCommandResponse from int to Status (enum).  If you are using the isOk() method then no change is required.
  * Created interface IXBee.java for XBee.java
  * Added all pins to XBeePins.java
  * Added abbreviated contructors for TX classes
  * Added isConnected method to XBee class
  * Added ZBForceSampleRequest class and corresponding example
  * Cleaned up examples
  * Added WPAN I/O Samples example
  * Added ZNet/ZB Pro I/O Samples example
  * Removed redundant synchronous send example

Version 0.5.1 (1/25/09)

This release fixes the following:

  * Source address on series 1 RX 64 bit packets was incorrectly set
  * Stackoverflow bug with the series 1 isD7On method.

See revisions 90-93 for line-by-line differences: http://code.google.com/p/xbee-api/source/list

Version 0.5 (1/17/09)

This release includes:

  * Support for Explicit TX/RX packets (0x11 and 0x91).  I have also included sample classes that demonstrate sending/receiving Explicit packets.

src/com/rapplogic/xbee/XBeePin.java
/**
 * Copyright (c) 2009 Andrew Rapp. All rights reserved.
 *  
 * This file is part of XBee-API.
 *  
 * XBee-API is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *  
 * XBee-API is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *  
 * You should have received a copy of the GNU General Public License
 * along with XBee-API.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.rapplogic.xbee;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

/**
 * Represents a configurable XBee pin and associated name, pin number, AT command, default capability and list of supported 
 * capabilities.
 * 
 * TODO add pin direction
 * TODO methods to filter list by Capability
 * 
 * @author andrew
 *
 */
public class XBeePin {

	private String name;
	private Integer pin;
	private String atCommand;
	private Integer atPin;
	
	private Capability defaultCapability;
	private String description;

	private List<Capability> capabilities = new ArrayList<Capability>();
	
	/**
	 * Contains all possible pin configurations and the associated AT command value
	 * 
	 * @author andrew
	 *
	 */
	public enum Capability {
		DISABLED (0),
		RTS_FLOW_CTRL(1),
		CTS_FLOW_CTRL(1),
		RSSI_PWM (1),
		ASSOC_LED (1),
		ANALOG_INPUT (2),
		PWM_OUTPUT (2),
		DIGITAL_INPUT (3),
		DIGITAL_OUTPUT_LOW (4),
		DIGITAL_OUTPUT_HIGH (5),
		UNMONITORED_INPUT (0), // only zigbee
		NODE_ID_ENABLED (1), // only zigbee
		RS485_TX_LOW(6), // only zigbee
		RS485_TX_HIGH(7); // only zigbee		
	
	    private final int value;
	    
	    Capability(int value) {
	        this.value = value;
	    }

		public int getValue() {
			return value;
		}
		
		public static String printAll(String delimiter) {
			StringBuffer sb = new StringBuffer();
			
			for (Capability cap : Capability.values()) {
				sb.append(System.getProperty("line.separator"));
				sb.append(delimiter);
				sb.append(cap);
				sb.append(delimiter);
				sb.append("0x" + cap.getValue());
				sb.append(cap.getValue());
				sb.append(delimiter);
			}
			
			return sb.toString();
		}
	}
	
	public XBeePin(String name, Integer pin, String atCommand, Integer atPin, Capability defaultCapability, String description, Capability... capabilityArr) {
		this.setName(name);
		this.setPin(pin);
		this.setAtCommand(atCommand);
		this.setAtPin(atPin);
		this.setDefaultCapability(defaultCapability);
		this.setDescription(description);
		
		if (capabilityArr != null) {
			for (Capability capability: capabilityArr) {
				this.getCapabilities().add(capability);
			}			
		}
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public Integer getPin() {
		return pin;
	}

	public void setPin(Integer pin) {
		this.pin = pin;
	}

	public String getAtCommand() {
		return atCommand;
	}

	public void setAtCommand(String atCommand) {
		this.atCommand = atCommand;
	}

	public Capability getDefaultCapability() {
		return defaultCapability;
	}

	public void setDefaultCapability(Capability defaultCapability) {
		this.defaultCapability = defaultCapability;
	}

	public String getDescription() {
		return description;
	}

	public void setDescription(String description) {
		this.description = description;
	}

	public List<Capability> getCapabilities() {
		return capabilities;
	}

	public void setCapabilities(List<Capability> capabilities) {
		this.capabilities = capabilities;
	}
	
	public Integer getAtPin() {
		return atPin;
	}

	public void setAtPin(Integer atPin) {
		this.atPin = atPin;
	}
	
	private final static List<XBeePin> zigBeePins = new ArrayList<XBeePin>();
	
	static {
		// notes: DIO13/DIO8/DIO9 not supported
		// TODO P0/P1 is missing pwm output option?? could it be 0x2?
		zigBeePins.add(new XBeePin("PWM0/RSSI/DIO10", 6, "P0", 10, Capability.RSSI_PWM, "PWM Output 0 / RX Signal Strength Indicator / Digital IO", 
				Capability.DISABLED, Capability.RSSI_PWM, Capability.DIGITAL_INPUT, Capability.DIGITAL_OUTPUT_LOW, 
				Capability.DIGITAL_OUTPUT_HIGH));
		zigBeePins.add(new XBeePin("PWM/DIO11", 7, "P1", 1, Capability.UNMONITORED_INPUT, "Digital I/O 11", Capability.UNMONITORED_INPUT, 
				Capability.DIGITAL_INPUT, Capability.DIGITAL_OUTPUT_LOW, Capability.DIGITAL_OUTPUT_HIGH));
		zigBeePins.add(new XBeePin("DIO12", 4, "P2", 2, Capability.UNMONITORED_INPUT, "Digital I/O 12", Capability.UNMONITORED_INPUT, 
				Capability.DIGITAL_INPUT, Capability.DIGITAL_OUTPUT_LOW, Capability.DIGITAL_OUTPUT_HIGH));
		zigBeePins.add(new XBeePin("AD0/DIO0/Commissioning Button", 20, "D0", 0, Capability.NODE_ID_ENABLED, 
				"Analog Input 0, Digital IO 0, or Commissioning Button", Capability.DISABLED, Capability.NODE_ID_ENABLED, 
				Capability.ANALOG_INPUT, Capability.DIGITAL_INPUT, Capability.DIGITAL_OUTPUT_LOW, Capability.DIGITAL_OUTPUT_HIGH));
		zigBeePins.add(new XBeePin("AD1/DIO1", 19, "D1", 1, Capability.DISABLED, "Analog Input 1 or Digital I/O 1", Capability.DISABLED, 
				Capability.ANALOG_INPUT, Capability.DIGITAL_INPUT, Capability.DIGITAL_OUTPUT_LOW, Capability.DIGITAL_OUTPUT_HIGH));
		zigBeePins.add(new XBeePin("AD2/DIO2", 18, "D2", 2, Capability.DISABLED, "Analog Input 2 or Digital I/O 2", Capability.DISABLED, 
				Capability.ANALOG_INPUT, Capability.DIGITAL_INPUT, Capability.DIGITAL_OUTPUT_LOW, Capability.DIGITAL_OUTPUT_HIGH));
		zigBeePins.add(new XBeePin("AD3/DIO3", 17, "D3", 3, Capability.DISABLED, "Analog Input 3 or Digital I/O 3", Capability.DISABLED, 
				Capability.ANALOG_INPUT, Capability.DIGITAL_INPUT, Capability.DIGITAL_OUTPUT_LOW, Capability.DIGITAL_OUTPUT_HIGH));
		zigBeePins.add(new XBeePin("DIO4", 11, "D4", 4, Capability.DISABLED, "Digital I/O 4", Capability.DISABLED, Capability.DIGITAL_INPUT, Capability.DIGITAL_OUTPUT_LOW, Capability.DIGITAL_OUTPUT_HIGH));
		zigBeePins.add(new XBeePin("Associate/DIO5", 15, "D5", 5, Capability.ASSOC_LED, "Associated Indicator, Digital I/O 5", 
				Capability.DISABLED, Capability.ASSOC_LED, Capability.DIGITAL_INPUT, Capability.DIGITAL_OUTPUT_LOW, 
				Capability.DIGITAL_OUTPUT_HIGH));
		zigBeePins.add(new XBeePin("CTS/DIO7", 12, "D7", 7, Capability.CTS_FLOW_CTRL, "Clear-to-Send Flow Control or Digital I/O 7", 
				Capability.DISABLED, Capability.CTS_FLOW_CTRL, Capability.DIGITAL_INPUT, Capability.DIGITAL_OUTPUT_LOW, 
				Capability.DIGITAL_OUTPUT_HIGH, Capability.RS485_TX_LOW, Capability.RS485_TX_HIGH));
		
		// TODO manual lists only RTS and disabled but x-ctu lists all digital capabilities
		zigBeePins.add(new XBeePin("RTS/DIO6", 16, "D6", 6, Capability.DISABLED, "Request-to-Send Flow Control, Digital I/O 6", 
				Capability.DISABLED, Capability.RTS_FLOW_CTRL, Capability.DIGITAL_INPUT, Capability.DIGITAL_OUTPUT_LOW, 
				Capability.DIGITAL_OUTPUT_HIGH));
		
		
		
		// other pins
		zigBeePins.add(new XBeePin("VCC", 1, null, null, null, "Power Supply", (Capability[])null));
		zigBeePins.add(new XBeePin("DOUT", 2, null, null, null, "UART Data Out", (Capability[])null));
		zigBeePins.add(new XBeePin("DIN", 3, null, null, null, "UART Data In", (Capability[])null));
		zigBeePins.add(new XBeePin("RESET", 5, null, null, null, "Module Reset (reset pulse must be at least 200 ns)", (Capability[])null));
		zigBeePins.add(new XBeePin("[reserved]", 8, null, null, null, "Do not connect", (Capability[])null));
		// DIO8 not supported according to manual
		zigBeePins.add(new XBeePin("DTR/SLEEP_RQ", 9, null, null, null, "Pin Sleep Control Line", (Capability[])null));
		zigBeePins.add(new XBeePin("GND", 10, null, null, null, "Ground", (Capability[])null));
		// DIO9 not supported according to manual
		zigBeePins.add(new XBeePin("ON/SLEEP", 13, null, null, null, "Module Status Indicator", (Capability[])null));
		zigBeePins.add(new XBeePin("VREF", 14, null, null, null, "Not used on this module. For compatibility with other XBee modules, we recommend connecting this pin to a voltage reference if Analog sampling is desired. Otherwise, connect to GND", (Capability[])null));
	
		Collections.sort(zigBeePins, new PinSorter());
	}
	
	/**
	 * Sorts by pin number
	 * 
	 * @author andrew
	 *
	 */
	private static class PinSorter implements Comparator {
		public int compare(Object o1, Object o2) {
			return ((XBeePin)o1).getPin().compareTo(((XBeePin)o2).getPin());
		}		
	}
	
	public static List<XBeePin> getZigBeePins() {
		return zigBeePins;
	}
	
	private final static List<XBeePin> wpanPins = new ArrayList<XBeePin>();
	
	static { 
		wpanPins.add(new XBeePin("DTR/SLEEP_RQ/DI8", 9, "D8", 8, Capability.DISABLED, "Pin Sleep Control Line or Digital Input 8", Capability.DISABLED, Capability.DIGITAL_INPUT));
		wpanPins.add(new XBeePin("CTS/DIO7", 12, "D7", 7, Capability.CTS_FLOW_CTRL, "Clear-to-Send Flow Control or Digital I/O 7", Capability.DISABLED, Capability.CTS_FLOW_CTRL, Capability.DIGITAL_INPUT, Capability.DIGITAL_OUTPUT_LOW, Capability.DIGITAL_OUTPUT_HIGH));
		wpanPins.add(new XBeePin("RTS/AD6/DIO6", 16, "D6", 6, Capability.DISABLED, "Request-to-Send Flow Control, Analog Input 6 or Digital I/O 6", Capability.DISABLED, Capability.RTS_FLOW_CTRL, Capability.DIGITAL_INPUT, Capability.DIGITAL_OUTPUT_LOW, Capability.DIGITAL_OUTPUT_HIGH));
		wpanPins.add(new XBeePin("Associate/AD5/DIO5", 15, "D5", 5, Capability.ASSOC_LED, "Associated Indicator, Analog Input 5 or Digital I/O 5", Capability.DISABLED, Capability.ASSOC_LED, Capability.ANALOG_INPUT, Capability.DIGITAL_INPUT, Capability.DIGITAL_OUTPUT_LOW, Capability.DIGITAL_OUTPUT_HIGH));
		wpanPins.add(new XBeePin("AD4/DIO4", 11, "D4", 4, Capability.DISABLED, "Analog Input 4 or Digital I/O 4", Capability.DISABLED, Capability.ANALOG_INPUT, Capability.DIGITAL_INPUT, Capability.DIGITAL_OUTPUT_LOW, Capability.DIGITAL_OUTPUT_HIGH));
		wpanPins.add(new XBeePin("AD3/DIO3", 17, "D3", 3, Capability.DISABLED, "Analog Input 3 or Digital I/O 3", Capability.DISABLED, Capability.ANALOG_INPUT, Capability.DIGITAL_INPUT, Capability.DIGITAL_OUTPUT_LOW, Capability.DIGITAL_OUTPUT_HIGH));
		wpanPins.add(new XBeePin("AD2/DIO2", 18, "D2", 2, Capability.DISABLED, "Analog Input 2 or Digital I/O 2", Capability.DISABLED, Capability.ANALOG_INPUT, Capability.DIGITAL_INPUT, Capability.DIGITAL_OUTPUT_LOW, Capability.DIGITAL_OUTPUT_HIGH));
		wpanPins.add(new XBeePin("AD1/DIO1", 19, "D1", 1, Capability.DISABLED, "Analog Input 1 or Digital I/O 1", Capability.DISABLED, Capability.ANALOG_INPUT, Capability.DIGITAL_INPUT, Capability.DIGITAL_OUTPUT_LOW, Capability.DIGITAL_OUTPUT_HIGH));
		wpanPins.add(new XBeePin("AD0/DIO0", 20, "D0", 0, Capability.DISABLED, "Analog Input 0 or Digital I/O 0", Capability.DISABLED, Capability.ANALOG_INPUT, Capability.DIGITAL_INPUT, Capability.DIGITAL_OUTPUT_LOW, Capability.DIGITAL_OUTPUT_HIGH));
		wpanPins.add(new XBeePin("PWM0/RSSI", 6, "P0", null, Capability.RSSI_PWM, "PWM Output 0 / RX Signal Strength Indicator", Capability.DISABLED, Capability.RSSI_PWM, Capability.PWM_OUTPUT));
		wpanPins.add(new XBeePin("PWM1", 7, "P1", null, Capability.DISABLED, "PWM Output 1", Capability.DISABLED,  Capability.RSSI_PWM, Capability.PWM_OUTPUT));
		
	
		// other pins
		wpanPins.add(new XBeePin("VCC", 1, null, null, null, "Power Supply", (Capability[])null));
		wpanPins.add(new XBeePin("DOUT", 2, null, null, null, "UART Data Out", (Capability[])null));
		wpanPins.add(new XBeePin("DIN/CONFIG", 3, null, null, null, "UART Data In", (Capability[])null));
		wpanPins.add(new XBeePin("DO8", 4, null, null, null, "Digital Output 8 (not supported as of 2/28/09)", (Capability[])null));
		wpanPins.add(new XBeePin("RESET", 5, null, null, null, "Module Reset (reset pulse must be at least 200 ns)", (Capability[])null));
		wpanPins.add(new XBeePin("[reserved]", 8, null, null, null, "Do not connect", (Capability[])null));
		wpanPins.add(new XBeePin("GND", 10, null, null, null, "Ground", (Capability[])null));
		wpanPins.add(new XBeePin("ON/SLEEP", 13, null, null, null, "Module Status Indicator", (Capability[])null));
		wpanPins.add(new XBeePin("VREF", 14, null, null, null, "Voltage Reference for A/D Inputs", (Capability[])null));
		
		Collections.sort(wpanPins, new PinSorter());
	}
	
	public static String printAll(List<XBeePin> pins, String delimiter) {
		
		StringBuffer sb = new StringBuffer();
		
		sb.append(delimiter);
		sb.append("*Pin #*");
		sb.append(delimiter);
		sb.append("*AT Command*");
		sb.append(delimiter);
		sb.append("*Description*");
		sb.append(delimiter);
		sb.append("*Default Configuration*");
		sb.append(delimiter);
		sb.append("*Supported Configurations (Value)*");
		sb.append(delimiter);
		sb.append("*Analog I/O Method*");
		sb.append(delimiter);
		sb.append("*Digital I/O Method*");
		sb.append(delimiter);
		
		for (XBeePin pin : pins) {
			sb.append(System.getProperty("line.separator"));
			
			sb.append(delimiter);
			sb.append(pin.getPin());
			sb.append(delimiter);
			sb.append(pin.getAtCommand() == null ? "n/a" : pin.getAtCommand());
			sb.append(delimiter);
			sb.append(pin.getDescription());
			sb.append(delimiter);
			sb.append(pin.getDefaultCapability() == null ? "n/a" : pin.getDefaultCapability());
			sb.append(delimiter);
			
			boolean first = false;
			
			for (Capability cap : pin.getCapabilities()) {
				if (!first) {
					first = true;
				} else {
					sb.append(", ");
				}
				
				sb.append(cap + " (0x" + cap.getValue() + ")");
				//sb.append(cap);
			}
		
			sb.append(delimiter);
			
			// analog methods
			if (pin.getCapabilities().contains(Capability.ANALOG_INPUT)) {
				sb.append("getAnalog" + pin.getAtPin() + "()");
			} else {
				sb.append("n/a");
			}

			sb.append(delimiter);
			
			// digital methods
			if (pin.getCapabilities().contains(Capability.DIGITAL_INPUT)) {
				sb.append("isD" + pin.getAtPin() + "On()");
			} else {
				sb.append("n/a");
			}
			
			sb.append(delimiter);
		}
		
		return sb.toString();
	}
	
	public static List<XBeePin> getWpanPins() {
		return wpanPins;
	}
	
	public static void main(String[] args) {
		// print for wiki table
		System.out.println(Capability.printAll("|| "));
		
		System.out.println("Series 1:");
		System.out.println(printAll(wpanPins, "|| "));
		System.out.println("Series 2:");
		System.out.println(printAll(zigBeePins, "|| "));
	}
}
